### PR TITLE
[TOOLS-2857] Scope chart data endpoint implemented - backend

### PR DESCRIPTION
**Pull request by emannuelalmeida**

### CODE REVIEW

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScopeScenarioService.java [hunk 1]
```diff
@@ -1,44 +1,229 @@
...
+                    ? details.getScopeConservative().doubleValue()
+                    : details.getScopeOptimistic().doubleValue();
+
+            currentScopeDone += weeklyThroughput;
+            projectionStep = currentScopeDone * ( 1 + factor.value() * scopeThreshold);
+
+            series.add(new ImmutablePair<>(currentProjectionDate, projectionStep));
+            currentProjectionDate = currentProjectionDate.plusWeeks(1);
+
+        };
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScopeScenarioService.java]: MINOR Remove this empty statement. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: strategical-dashboard/src/test/java/objective/taskboard/monitor/ScopeScenarioServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+           .scopeDoneDataSeriesIs()
+                .at("2019-09-30", 0.0)
+                .at("2019-10-07", 30.0)
+                .at("2019-10-14", 60.0)
+           .conservativeProjectionDataSeriesIs()
+                .at("2019-10-21", 81.0)
+           .optimisticProjectionDataSeriesIs()
+                .at("2019-10-21", 99.0)
+           .scopeTopOnTrackDataSeriesIs()
+                .at("2019-10-14", 52.8)
```
Comment (taksan) [inline strategical-dashboard/src/test/java/objective/taskboard/monitor/ScopeScenarioServiceTest.java]: This is testing only one date, but the implementation generates all dates until current date.
Also, the implementation doesn't generate date for future dates, but the chart in slide 25 has data until the delivery date.
This applies to next series as well.

File: strategical-dashboard/src/test/java/objective/taskboard/monitor/ScopeScenarioServiceDSL.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+        public ScopeScenarioServiceAsserter at(String date) {
+            LocalDate parsedDate = parseLocalDate(date);
+            LocalDate lastDate = listToAssert.get(listToAssert.size() -1).getLeft();
+            assertEquals(parsedDate, lastDate);
+
+            return this;
+        }
+
+        private double getMin(double valueToAssert, double totalScope) {
```
Comment (taksan) [inline strategical-dashboard/src/test/java/objective/taskboard/monitor/ScopeScenarioServiceDSL.java]: This method is not being used

File: strategical-dashboard/src/test/java/objective/taskboard/monitor/ScopeScenarioServiceDSL.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+    private FollowUpSnapshotService followUpSnapshotService = mock(FollowUpSnapshotService.class);
+    private ScenarioFacade scenarioFacade = mock(ScenarioFacade.class);
+    private FixedClock clock = new FixedClock();
+    private List<ScenarioDetailsDto> scenarioDetailsDtos = new ArrayList<>();
+    private ProjectFilterConfiguration project = mock(ProjectFilterConfiguration.class);
+    private String projectKey = "PROJECT";
+    private ZoneId timezone = ZoneId.systemDefault();
+    private ScopeChartData result;
+    private double totalScope;
```
Comment (taksan) [inline strategical-dashboard/src/test/java/objective/taskboard/monitor/ScopeScenarioServiceDSL.java]: This variable is not being used

File: core/src/main/java/objective/taskboard/utils/DateTimeUtils.java [hunk 5]
```diff
@@ -250,4 +250,8 @@ public static double secondsToHours(double seconds) {
     public static long hoursToSeconds(double hours) {
         return (long) (hours * DateTimeUtils.HOUR_IN_SECONDS);
     }
+    
+    public static LocalDate max(LocalDate dateA, LocalDate dateB) {
+        return Stream.of(dateA, dateB).max(LocalDate::compareTo).get();
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/utils/DateTimeUtils.java]: MAJOR Call "Optional#isPresent()" before accessing the value. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3655)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScopeScenarioService.java [hunk 6]
```diff
@@ -1,44 +1,269 @@
...
+
+            return thresholdPoint;
+        }).collect(toList());
+    }
+    
+
+    private List<LocalDate> weekListBetween(LocalDate startDate, LocalDate endDate) {
+        long projectWeekCount = startDate.until(endDate, ChronoUnit.WEEKS);
+        return LongStream.range(0, projectWeekCount)
+                .mapToObj(weekNum -> startDate.plusWeeks(weekNum))
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScopeScenarioService.java]: MINOR Replace this lambda with a method reference. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1612)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**